monai/data/dataset.py

```diff
@@ -838,16 +838,15 @@ def _compute_cache_num(data_len: int):
 
         if self.runtime_cache in (False, None):  # prepare cache content immediately
             self._cache = self._fill_cache(indices)
-            return
-        if isinstance(self.runtime_cache, str) and "process" in self.runtime_cache:
+        elif isinstance(self.runtime_cache, str) and "process" in self.runtime_cache:
             # this must be in the main process, not in dataloader's workers
             self._cache = Manager().list([None] * self.cache_num)
-            return
-        if (self.runtime_cache is True) or (isinstance(self.runtime_cache, str) and "thread" in self.runtime_cache):
+        elif (self.runtime_cache is True) or (isinstance(self.runtime_cache, str) and "thread" in self.runtime_cache):
             self._cache = [None] * self.cache_num
-            return
-        self._cache = self.runtime_cache  # type: ignore
-        return
+        else:
+            self._cache = self.runtime_cache  # type: ignore
+            if len(self._cache) == 0:
+                self._cache[:] = [None] * self.cache_num
```

Comment:
I think you should just use len(data) outside of this class to prepare the cache. Given the hashing mechanism, the actual length is not known anyway. But it'll be at most len(data).

Reply:
unfortunately, a user may not know the len(data) at the moment of creation of ListProxy (if it's created in master process as  Manager().list()) only in the child processes when the datalist is available, we know len(data).  it will be unnecessary extra coding on a user-side to mitigate it.

and if len(self._cache) is different from self.cache_num, it can create unforeseen bugs in the future.   Currently it'll probably crash if len(self._cach)==0, and self.cache_num>0.

Reply:
the runtime_cache and `data: Sequence` must be both provided when creating the instance, why is it not available?  the `len` method is not necessarily implemented by the user-provided object and indeed is not required by the cache reading logic.

Reply:
1) there reason why len(data) is not available is because a master process needs to create a shared list before anything else

```
def subfunction(rank, shared_list):
  #mainlogic to load data
  data = ...
  CacheData(data=data, runtime_cache=shared_list)

shared_list=Manager().list()
np.swawn(subfunction, nproc=8, args(shared_list))
```

2)   in your definition of `runtime_cache: Union[bool, str, List, ListProxy] = False,`,  so at the moment when len(self._cache) is called it can only be List or ListProxy.  we can also check for isinstance(ListProxy) if you'd like

Reply:
Within the subfunction, the length of the shared list can be changed, right? I mean if we make less assumptions, the user will have more flexibility

Reply:
I've added to check for type isinstance(self._cache, (list, ListProxy))

Reply:
> Within the subfunction, the length of the shared list can be changed, right? I mean if we make less assumptions, the user will have more flexibility

we need to simplify things for a user, it was easy to use it before this PR https://github.com/Project-MONAI/MONAI/pull/5630

now as it stands,  a user needs to remember to initialize it's own cache, and make sure it's length is correct  (and no shorter than num_cache) otherwise it will crash. which is very inconvenient.

if nothing else, this PR does not hurt anything, and avoid the crash if empty list is provided.  is there any reason why not merge it?